pkg/commands/root.go

```diff
@@ -1,160 +1,168 @@
 package commands
 
 import (
+	"errors"
 	"fmt"
 	"os"
-	"runtime"
-	"runtime/pprof"
-	"runtime/trace"
-	"strconv"
+	"slices"
 
+	"github.com/fatih/color"
 	"github.com/spf13/cobra"
 	"github.com/spf13/pflag"
 
 	"github.com/golangci/golangci-lint/pkg/config"
-	"github.com/golangci/golangci-lint/pkg/exitcodes"
 	"github.com/golangci/golangci-lint/pkg/logutils"
+	"github.com/golangci/golangci-lint/pkg/report"
 )
 
-const (
-	// envHelpRun value: "1".
-	envHelpRun        = "HELP_RUN"
-	envMemProfileRate = "GL_MEM_PROFILE_RATE"
-)
+func Execute(info BuildInfo) error {
+	return newRootCommand(info).Execute()
+}
+
+type rootOptions struct {
+	PrintVersion bool
+
+	Verbose bool   // Persistent flag.
+	Color   string // Persistent flag.
+}
+
+type rootCommand struct {
+	cmd  *cobra.Command
+	opts rootOptions
+
+	log logutils.Log
+}
+
+func newRootCommand(info BuildInfo) *rootCommand {
+	c := &rootCommand{}
 
-func (e *Executor) initRoot() {
 	rootCmd := &cobra.Command{
 		Use:   "golangci-lint",
 		Short: "golangci-lint is a smart linters runner.",
 		Long:  `Smart, fast linters runner.`,
 		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, _ []string) error {
+			if c.opts.PrintVersion {
+				_ = printVersion(logutils.StdOut, info)
+				return nil
+			}
+
 			return cmd.Help()
 		},
-		PersistentPreRunE:  e.persistentPreRun,
-		PersistentPostRunE: e.persistentPostRun,
 	}
 
-	initRootFlagSet(rootCmd.PersistentFlags(), e.cfg)
-
-	e.rootCmd = rootCmd
-}
+	fs := rootCmd.Flags()
+	fs.BoolVar(&c.opts.PrintVersion, "version", false, wh("Print version"))
 
-func (e *Executor) persistentPreRun(_ *cobra.Command, _ []string) error {
-	if e.cfg.Run.PrintVersion {
-		_ = printVersion(logutils.StdOut, e.buildInfo)
-		os.Exit(exitcodes.Success) // a return nil is not enough to stop the process because we are inside the `preRun`.
-	}
+	setupRootPersistentFlags(rootCmd.PersistentFlags(), &c.opts)
```

Comment:
Ah, here it's pretty clear what persistent flags means, maybe the comments I commented about can be removed since they add more confusion than clarification.